File: Backend/src/services/weather.py
from repositories.weather import WeatherRepository


class WeatherService:
    def __init__(self, repository: WeatherRepository | None = None):
        self.repository = repository or WeatherRepository()

    async def get_current_weather(
        self,
        lat: float | None = None,
        lon: float | None = None,
        city: str | None = None,
    ):
        return await self.repository.get_current_weather(lat=lat, lon=lon, city=city)

    async def get_forecast(
        self,
        days: int = 5,
        lat: float | None = None,
        lon: float | None = None,
        city: str | None = None,
    ):
        return await self.repository.get_forecast(lat=lat, lon=lon, days=days, city=city)

    async def geocode_city(self, city: str):
        return await self.repository.geocode_city(city=city)

File: Frontend/nimrag-frontend/src/services/weather.ts
import { getJson } from './api'
import type { WeatherCurrentResponse, WeatherForecastResponse } from '../types/weather'

export interface WeatherQuery {
  lat?: number
  lon?: number
  city?: string
}

function buildWeatherParams(query: WeatherQuery): URLSearchParams {
  const p = new URLSearchParams()
  if (query.lat !== undefined) p.set('lat', String(query.lat))
  if (query.lon !== undefined) p.set('lon', String(query.lon))
  if (query.city) p.set('city', query.city)
  return p
}

export function getCurrentWeather(query: WeatherQuery = {}): Promise<WeatherCurrentResponse> {
  const qs = buildWeatherParams(query).toString()
  return getJson<WeatherCurrentResponse>(qs ? `/weather/current?${qs}` : '/weather/current')
}

export function getForecast(days = 5, query: WeatherQuery = {}): Promise<WeatherForecastResponse> {
  const p = buildWeatherParams(query)
  p.set('days', String(days))
  return getJson<WeatherForecastResponse>(`/weather/forecast?${p.toString()}`)
}